flow/Trace.cpp

```diff
@@ -779,66 +783,18 @@ TraceEvent& TraceEvent::detailImpl( std::string&& key, std::string&& value, bool
 	return *this;
 }
 
-TraceEvent& TraceEvent::detail( std::string key, std::string value ) {
-	return detailImpl(std::move(key), std::move(value));
-}
-TraceEvent& TraceEvent::detail( std::string key, double value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), value);
-	return detailfNoMetric( std::move(key), "%g", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, int value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%d", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, unsigned value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%u", value );
+void TraceEvent::setField(const char* key, int64_t value) {
+	tmpEventMetric->setField(key, value);
 }
-TraceEvent& TraceEvent::detail( std::string key, long int value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%ld", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, long unsigned int value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%lu", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, long long int value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%lld", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, long long unsigned int value ) {
-	init();
-	if(enabled)
-		tmpEventMetric->setField(key.c_str(), (int64_t)value);
-	return detailfNoMetric( std::move(key), "%llu", value );
-}
-TraceEvent& TraceEvent::detail( std::string key, const NetworkAddress& value ) {
-	return detailImpl( std::move(key), value.toString() );
-}
-TraceEvent& TraceEvent::detail( std::string key, const IPAddress& value ) {
-	return detailImpl( std::move(key), value.toString() );
-}
-TraceEvent& TraceEvent::detail( std::string key, const UID& value ) {
-	return detailf( std::move(key), "%016llx", value.first() );  // SOMEDAY: Log entire value?  We also do this explicitly in some "lists" in various individual TraceEvent calls
-}
-TraceEvent& TraceEvent::detailext( std::string key, StringRef const& value ) {
-	return detailImpl(std::move(key), value.printable());
```

Comment:
I think there is a behavior change here: a call to `detailext` before would call `printable`, but it looks like some calls to `detailext` were replaced with something that does not call `printable`.

Reply:
The purpose of this is to get rid of printable - so this is intended. Instead this functionality is implemented in the `Traceable` concept

Reply:
Does that mean we're calling printable somewhere else for StringRef and (and similar for Optional<Standalone<StringRef>>)? Where is that happening?

Reply:
Yes. The way this works is that `Traceable<T>` implements a static method `static std::string toString(const T& value)` which gets called only if this string is actually needed (so if  `TraceEvent::enabled == false`, it won't be called at all). For strings I tried to optimize this as good as possible:

- In a first iteration it counts the number of characters that need escaping.
- If the string doesn't need any escaping, it is simply returned. For `std::string` this uses perfect forwarding, so depending on how `detail` is called we can get away without any allocations. (For `StringRef` we have to copy)
- If the string needs formatting, we can now calculate the exact length of the resulting string. This means that only one allocation is needed. Then the string is copied and escaped in one go.

For compound types (like `Optional<Standalone<StringRef>>`) the whole thing is defined recursively. There's a specialization for each type. So `Traceable<Optional<T>>::value` is true iff `Traceable<T>::value` is true. So the implementation of `Traceable<Optional<T>>` would look something like this:

```
template<class T>
struct Traceable<Optional<T> {
    static constexpr bool value = Traceable<T>::value;

    static std::string toString(const Optional<T>& value) {
        if (value.present()) return "[notset]";
        return Traceable<T>::toString(value.get());
    }
};
```

Reply:
Would it make sense to defer the work of ensuring we're emitting valid json/xml to {Json/Xml}TraceLogFormatter? Currently those have functions to escape strings so I think we're escaping twice

Reply:
yes, currently we do escaping twice. I think we should change that but I think this should happen in another pull request. I don't want to bloat this much more.

Reply:
I gather from the thumbs up that this thread is resolved.